src/axom/core/FlatMapUtil.hpp

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) 2017-2025, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#ifndef Axom_Core_FlatMap_Util_HPP
+#define Axom_Core_FlatMap_Util_HPP
+
+#include "axom/config.hpp"
+#include "axom/core/FlatMap.hpp"
+#include "axom/core/execution/reductions.hpp"
+
+namespace axom
+{
+namespace detail
+{
+
+struct SpinLock
+{
+  int value {0};
+
+  AXOM_HOST_DEVICE bool tryLock()
+  {
+    int still_locked = 0;
+#if defined(__HIP_DEVICE_COMPILE__)
+    still_locked = __hip_atomic_exchange(&value, 1, __ATOMIC_ACQUIRE, __HIP_MEMORY_SCOPE_AGENT);
+#elif defined(AXOM_USE_RAJA) && defined(__CUDA_ARCH__)
+    still_locked = RAJA::atomicExchange<RAJA::cuda_atomic>(&value, 1);
+    // We really want an acquire-fenced atomic here
+    __threadfence();
+#elif defined(AXOM_USE_RAJA) && defined(AXOM_USE_OPENMP)
+    still_locked = RAJA::atomicExchange<RAJA::omp_atomic>(&value, 1);
+    std::atomic_thread_fence(std::memory_order_acquire);
+#endif
+    return !still_locked;
+  }
+
+  AXOM_HOST_DEVICE void unlock()
+  {
+#if defined(__HIP_DEVICE_COMPILE__)
+    __hip_atomic_exchange(&value, 0, __ATOMIC_RELEASE, __HIP_MEMORY_SCOPE_AGENT);
+#elif defined(AXOM_USE_RAJA) && defined(__CUDA_ARCH__)
+    // We really want a release-fenced atomic here
+    __threadfence();
+    RAJA::atomicExchange<RAJA::cuda_atomic>(&value, 0);
+#elif defined(AXOM_USE_RAJA) && defined(AXOM_USE_OPENMP)
+    std::atomic_thread_fence(std::memory_order_release);
+    RAJA::atomicExchange<RAJA::omp_atomic>(&value, 0);
+#else
+    value = 0;
+#endif
+  }
+};
+
+}  // namespace detail
+
+template <typename KeyType, typename ValueType, typename Hash>
+template <typename ExecSpace>
+auto FlatMap<KeyType, ValueType, Hash>::create(ArrayView<KeyType> keys,
+                                               ArrayView<ValueType> values,
+                                               Allocator allocator) -> FlatMap
+{
+  assert(keys.size() == values.size());
+
+  IndexType num_elems = keys.size();
```

Comment:
Thanks for all the comments!

Suggestion:

This code is fairly complex. Marking const variables as `const` would help a bit w/ readability/understandability/correctness:

```suggestion
  const IndexType num_elems = keys.size();
```